ball.cpp

```diff
@@ -1,40 +1,80 @@
 #include "ball.h"
 
+#include "opponent.h"
+#include "player.h"
 #include "rectangle.h"
 
 namespace pong
 {
 
 static constexpr float BALL_WIDTH = 25.0f;
-static constexpr float BALL_SPEED = 4.5f;
-static int id = 0;
+static constexpr float BALL_SPEED_BOUNCE_INCREMENT = 0.25f;
+static constexpr float Y_STARTING_POSITION_BOUNDS = 275.0f;
 
 void Ball::OnStart()
 {
     mMesh = std::make_unique<Rectangle>(BALL_WIDTH, BALL_WIDTH);
     mColliderBox = std::make_unique<ColliderBox>(BALL_WIDTH, BALL_WIDTH);
     SetInstanceName("Ball");
-    mXSpeed = BALL_SPEED;
-    mYSpeed = BALL_SPEED;
+    ResetBall();
 }
 
 void Ball::OnUpdate()
 {
-    const glm::vec3 newPosition = GetPosition() + glm::vec3(mXSpeed, mYSpeed, 0.0f);
+    const glm::vec3 newPosition = GetPosition() + mVelocity;
     SetPosition(newPosition);
 }
 
 void Ball::OnCollisionStart(GameObject& other)
 {
-    if (other.GetInstanceName() == "Player" || other.GetInstanceName() == "Opponent"
-     || other.GetInstanceName() == "WallV")
+    if (other.GetInstanceName() == "ScoreArea")
     {
-        mXSpeed = -mXSpeed;
+        ResetBall();
+    }
+    else if (other.GetInstanceName() == "Player" || other.GetInstanceName() == "Opponent")
+    {
+        mSpeed += BALL_SPEED_BOUNCE_INCREMENT;
+
+        const float yPosDiff = GetPosition().y - other.GetPosition().y;
+        const float paddleHeight = other.GetColliderBox()->GetHeight();
+        const float pointInHeight = paddleHeight / 2 - abs(yPosDiff);
+
+        float percent = abs(pointInHeight / (paddleHeight / 2));
```

Comment:
could consider absf, most math things have a double version and a float version

aka

sin works on doubles
sinf works on floats